### PR TITLE
AZP: CUDA compatible

### PR DESCRIPTION
**Pull request by avildema**
launch test on official CUDA images

### CODE REVIEW

File: buildlib/azure-pipelines-pr.yml [hunk 1]
```diff
@@ -61,6 +61,67 @@ resources:
       image: rdmz-harbor.rdmz.labs.mlnx/swx-infra/x86_64/sles12sp5/builder:mofed-5.0-1.0.0.0
       options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
 
+    - container: centos7_cuda_11_0
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:64]: can we put this huge list in a separate YML file?

File: buildlib/azure-pipelines-pr.yml [hunk 2]
```diff
@@ -340,3 +401,96 @@ stages:
         parameters:
             name: gpu
             demands: ucx_gpu -equals yes
+
+  - stage: Cuda
+    dependsOn: [Static_check]
+    jobs:
+      - job: autogen_ucx
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:408]: job: source_tarball

File: buildlib/tools/test_malloc_hook.sh [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	then
+		azure_log_error "Cuda not found"
+		exit 1
+	fi
+}
+
+build() {
+	../configure --enable-gtest --enable-examples --with-cuda=/usr/local/cuda --prefix=$ucx_inst
+
+	make
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:42]: use -j`nproc` ?

File: buildlib/tools/test_malloc_hook.sh [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	make install
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:43]: no need to install

File: buildlib/tools/test_malloc_hook.sh [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+test_malloc_hook() {
+	echo "==== Running malloc hooks test ===="
+	if [ -x ./test/apps/test_tcmalloc ]
+	then
+		./test/apps/test_tcmalloc
+	fi
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:51]: can remove this

File: buildlib/tools/test_malloc_hook.sh [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+	if [ "X$have_cuda" == "Xyes" ]
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:53]: why need?
since try_load_cuda_env exits if hava_cuda==no

File: buildlib/tools/test_malloc_hook.sh [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+			unset UCX_MEM_CUDA_HOOK_MODE
+		done
+	fi
+}
+
+prepare
+build
+try_load_cuda_env
+test_malloc_hook
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:98]: add spaceline

File: buildlib/tools/test_malloc_hook.sh [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	have_cuda=no
+	if [ -f "/proc/driver/nvidia/version" ]; then
+		have_cuda=yes
+	fi
+	if [ "X$have_cuda" == "Xno" ]
+	then
+		azure_log_error "Cuda not found"
+		exit 1
+	fi
+}
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:37]: 1. rename the function to check_cuda_driver
2. can simplify: if "/proc/driver/nvidia/version" not found, exit with error

File: buildlib/tools/test_malloc_hook.sh [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+prepare() {
+	echo " ==== Prepare ===="
+	env
+	cd ${WORKSPACE}
+	if [ -d build-test ]
+	then
+		chmod u+rwx build-test -R
+		rm -rf build-test
+	fi
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:19]: this dir should not exist since we're running in docker

**Review by yosefe — COMMENTED** [review]
in general can simplify this script since some parts are copy/paste from test_jenkins.sh but not needed here

**Comment by yosefe** [comment]
Azure job did not run:
image

File: buildlib/azure-pipelines-pr.yml (hunk 1 above)
Comment (avildema) [reply buildlib/azure-pipelines-pr.yml]: it's impossible 
https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops
https://docs.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema

resources doesn't support templates
and extend: template is not applicable 

we can leave it here and for example move all stages to separate files ( like refactoring in other PR)

Comment (yosefe) [reply buildlib/azure-pipelines-pr.yml:64]: ok

**Review by yosefe — APPROVED** [review]

File: buildlib/tools/test_malloc_hook.sh [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+build() {
+	../configure --enable-gtest --enable-examples --with-cuda=/usr/local/cuda --prefix=$ucx_inst
+
+	make -j$(nproc)
+}
+
+test_malloc_hook() {
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:35]: pls remove this test form test_jenkins.sh

File: buildlib/pr/cuda.yml [hunk 11]
```diff
@@ -0,0 +1,89 @@
+jobs:
+  - job: source_tarball
+    displayName: Test autogen_cuda
```
Comment (yosefe) [inline buildlib/pr/cuda.yml:3]: Create UCX source tarball

File: buildlib/pr/cuda.yml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      demands:
+        - ucx_docker -equals yes
+    container: centos7
+
+    steps:
+      - checkout: self
+        clean: true
+        fetchDepth: 100
+      - bash: |
+          set -eE
```
Comment (yosefe) [inline buildlib/pr/cuda.yml:15]: set -xeE

File: buildlib/pr/main.yml [hunk 13]
```diff
@@ -121,8 +194,9 @@ stages:
     - template: tests.yml
       parameters:
         name: gpu
-        demands: ucx_gpu -equals yes
+        demands: ucx_gpu_test -equals yes
```
Comment (yosefe) [inline buildlib/pr/main.yml:197]: should we remove this now? seems it runs on both docker and bare metal?

File: buildlib/tools/test_malloc_hook.sh [hunk 14]
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash -eExl
```
Comment (yosefe) [inline buildlib/tools/test_malloc_hook.sh:1]: move this file under pr/.
maybe move all cuda files under buildlib/pr/cuda ?

**Review by yosefe — COMMENTED** [review]
seems RDMA transports are not found in the container

File: buildlib/pr/main.yml (hunk 13 above)
Comment (avildema) [reply buildlib/pr/main.yml:194]: yes, it need for test, I have one agent with one label `ucx_gpu_test` and this label always free

**Comment by avildema** [comment]
@yosefe Could you please check result for test_jenkins.sh for GPU on docker images 
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=26010&view=logs&j=466d7b3b-a401-53bf-628f-285bcb9960ef&t=1629ddf3-f393-545f-bef9-c35a9a07dc48

**Comment by avildema** [comment]
bot:retest

File: buildlib/pr/cuda/test_malloc_hook.sh [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+#
+check_cuda_driver() {
+	if [ ! -f "/proc/driver/nvidia/version" ]; then
+		azure_log_error "Cuda driver not found"
+		exit 1
+	fi
+}
+
+build() {
+	../configure --enable-gtest --enable-examples --with-cuda=/usr/local/cuda --prefix=$ucx_inst
```
Comment (yosefe) [inline buildlib/pr/cuda/test_malloc_hook.sh:30]: use configure-devel

File: buildlib/pr/main.yml [hunk 16]
```diff
@@ -1,8 +1,11 @@
 resources:
   containers:
     - container: centos7
-      image: rdmz-harbor.rdmz.labs.mlnx/ucx/centos7:3
+      image: rdmz-harbor.rdmz.labs.mlnx/ucx/centos7:4
       options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
+    - container: centos7_cuda11
+      image: rdmz-harbor.rdmz.labs.mlnx/ucx/centos7-mofed5.4-cuda11:1
+      options: --net=host --gpus all --ulimit memlock=-1:-1 -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools --device=/dev/infiniband/
```
Comment (yosefe) [inline buildlib/pr/main.yml:8]: 1. --net=host works?
2. why need `--ulimit memlock=-1:-1 -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools --device=/dev/infiniband/` only for this container and not all other as well?

File: buildlib/pr/main.yml (hunk 13 above)
Comment (yosefe) [reply buildlib/pr/main.yml:201]: still needed?

File: buildlib/pr/main.yml [hunk 17]
```diff
@@ -135,8 +198,9 @@ stages:
     - template: tests.yml
       parameters:
         name: gpu
-        demands: ucx_gpu -equals yes
+        demands: ucx_gpu_test -equals yes
         test_perf: 1
+        container: centos7_cuda11
```
Comment (yosefe) [inline buildlib/pr/main.yml:203]: so we run test_jenkins.sh only on centos7_cuda11, but test_memhooks on all matrix?

File: buildlib/pr/cuda/cuda.yml [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        ubuntu18_cuda_11_0:
+          CONTAINER: ubuntu18_cuda_11_0
+        ubuntu18_cuda_11_1:
+          CONTAINER: ubuntu18_cuda_11_1
+        ubuntu18_cuda_11_2:
+          CONTAINER: ubuntu18_cuda_11_2
+        ubuntu18_cuda_11_3:
+          CONTAINER: ubuntu18_cuda_11_3
+        ubuntu18_cuda_11_4:
+          CONTAINER: ubuntu18_cuda_11_4
```
Comment (yosefe) [inline buildlib/pr/cuda/cuda.yml:61]: pls add cuda 11.5 and 11.6

File: buildlib/pr/main.yml (hunk 16 above)
Comment (avildema) [reply buildlib/pr/main.yml:8]: 1 - No, we have hack agents
2 -  it need for same test, we cannot add different options for different stages

Comment (yosefe) [reply buildlib/pr/main.yml:8]: 1- so can remove it?
2 - question is why not add these parameters to all cuda dockers below (btw is it possible to make it templated some how to not repeat the same options for every container) ?

File: buildlib/pr/main.yml (hunk 17 above)
Comment (avildema) [reply buildlib/pr/main.yml:203]: for all cuda official docker images

File: buildlib/pr/main.yml (hunk 16 above)
Comment (avildema) [reply buildlib/pr/main.yml:8]: 1 - No , it use in hack script, if script find --net=host, script remove --net=<azure_net>
2 - will try via variable, maybe azure will not fails

File: buildlib/pr/main.yml [hunk 19]
```diff
@@ -1,121 +1,149 @@
+variables:
+  DOCKER_OPT_SHARES: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
```
Comment (yosefe) [inline buildlib/pr/main.yml:2]: DOCKER_OPT_VOLUMES

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@avildema pls squash

**Review by yosefe — APPROVED** [review]

**Pull request merged**